File: sdk/python/kraken/identity.py
"""Identity management operations (SOUL.md, user model, AGENTS.md, identity links)."""

from __future__ import annotations

from kraken._transport import Transport
from kraken.models import AgentsMd, IdentityLink, Soul, UserModel


class Identity:
    """Manage the agent's identity (SOUL.md), user model, AGENTS.md, and identity links."""

    def __init__(self, transport: Transport) -> None:
        self._t = transport

    def get_soul(self) -> Soul:
        """Get the current SOUL.md content."""
        data = self._t.get("/v1/identity/soul")
        return Soul.model_validate(data)

    def set_soul(self, content: str) -> Soul:
        """Update the SOUL.md content (agent personality)."""
        data = self._t.put("/v1/identity/soul", json={"content": content})
        return Soul.model_validate(data)

    def get_user_model(self) -> UserModel:
        """Get the auto-maintained user model."""
        data = self._t.get("/v1/identity/user-model")
        return UserModel.model_validate(data)

    def get_agents_md(self) -> AgentsMd:
        """Get the AGENTS.md project context."""
        data = self._t.get("/v1/identity/agents-md")
        return AgentsMd.model_validate(data)

    def set_agents_md(self, content: str) -> AgentsMd:
        """Update the AGENTS.md project context."""
        data = self._t.put("/v1/identity/agents-md", json={"content": content})
        return AgentsMd.model_validate(data)

    def link_identity(
        self,
        canonical_user_id: str,
        provider: str,
        provider_user_id: str,
        display_name: str | None = None,
    ) -> IdentityLink:
        """Link a platform identity to a canonical user ID."""
        payload: dict = {
            "canonical_user_id": canonical_user_id,
            "provider": provider,
            "provider_user_id": provider_user_id,
        }
        if display_name is not None:
            payload["display_name"] = display_name
        data = self._t.post("/v1/identity/links", json=payload)
        return IdentityLink.model_validate(data)

    def list_identity_links(
        self, canonical_user_id: str | None = None
    ) -> list[IdentityLink]:
        """List identity links, optionally filtered by canonical user ID."""
        params = {}
        if canonical_user_id:
            params["canonical_user_id"] = canonical_user_id
        data = self._t.get("/v1/identity/links", params=params)
        return [IdentityLink.model_validate(link) for link in data.get("links", [])]

File: server/src/api/identity.ts
import { Hono } from "hono";
import { z } from "zod";
import { zValidator } from "@hono/zod-validator";
import {
  getSoul,
  getUserModel,
  setSoul,
  getAgentsMd,
  setAgentsMd,
  linkIdentity,
  listIdentityLinks,
} from "../services/identity.js";
import { config } from "../config.js";

export const identityRouter = new Hono();

// --- Schemas ---
const updateSoulSchema = z.object({
  content: z.string().min(1).max(config.KRAKEN_MAX_SOUL_CHARS),
});

const updateAgentsMdSchema = z.object({
  content: z.string().min(1).max(4000),
});

const linkIdentitySchema = z.object({
  canonical_user_id: z.string().min(1),
  provider: z.string().min(1),
  provider_user_id: z.string().min(1),
  display_name: z.string().optional(),
});

// --- GET /v1/identity/soul ---
identityRouter.get("/soul", async (c) => {
  const soul = await getSoul();
  return c.json({ content: soul.content, updated_at: soul.updatedAt });
});

// --- PUT /v1/identity/soul ---
identityRouter.put(
  "/soul",
  zValidator("json", updateSoulSchema),
  async (c) => {
    const body = c.req.valid("json");
    const soul = await setSoul(body.content);
    return c.json({ content: soul.content, updated_at: soul.updatedAt });
  },
);

// --- GET /v1/identity/user-model ---
identityRouter.get("/user-model", async (c) => {
  const userModel = await getUserModel();
  return c.json({ content: userModel.content, updated_at: userModel.updatedAt });
});

// --- GET /v1/identity/agents-md ---
identityRouter.get("/agents-md", async (c) => {
  const agentsMd = await getAgentsMd();
  return c.json({ content: agentsMd.content, updated_at: agentsMd.updatedAt });
});

// --- PUT /v1/identity/agents-md ---
identityRouter.put(
  "/agents-md",
  zValidator("json", updateAgentsMdSchema),
  async (c) => {
    const body = c.req.valid("json");
    const agentsMd = await setAgentsMd(body.content);
    return c.json({ content: agentsMd.content, updated_at: agentsMd.updatedAt });
  },
);

// --- POST /v1/identity/links ---
identityRouter.post(
  "/links",
  zValidator("json", linkIdentitySchema),
  async (c) => {
    const body = c.req.valid("json");
    const link = await linkIdentity({
      canonicalUserId: body.canonical_user_id,
      provider: body.provider,
      providerUserId: body.provider_user_id,
      displayName: body.display_name,
    });
    return c.json(
      {
        id: link.id,
        canonical_user_id: link.canonicalUserId,
        provider: link.provider,
        provider_user_id: link.providerUserId,
        display_name: link.displayName,
      },
      201,
    );
  },
);

// --- GET /v1/identity/links ---
identityRouter.get("/links", async (c) => {
  const canonicalUserId = c.req.query("canonical_user_id");
  const links = await listIdentityLinks(canonicalUserId);
  return c.json({ links });
});
